pkg/server/share_manager.go

```diff
@@ -218,10 +220,36 @@ func (m *ShareManager) runHealthCheck() {
 	}
 }
 
-func (m *ShareManager) hasHealthyVolume() bool {
+func (m *ShareManager) hasHealthyVolume(vol volume.Volume, devicePath string) bool {
 	mountPath := types.GetMountPath(m.volume.Name)
-	err := exec.CommandContext(m.context, "ls", mountPath).Run()
-	return err == nil
+	if err := exec.CommandContext(m.context, "ls", mountPath).Run(); err != nil {
+		return false
+	}
+
+	mounter := mount.New("")
+	mountPoints, _ := mounter.List()
+	for _, mp := range mountPoints {
+		if mp.Path == mountPath && isMountPointReadOnly(mp) {
+			m.logger.Warn("Remounting the volume since mount point become read only")
+
+			cmd := exec.CommandContext(m.context, "mount", "-o", "remount,rw", mountPath)
+			if out, err := cmd.CombinedOutput(); err != nil {
+				m.logger.WithError(err).Errorf("remount failed with output: %s", out)
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isMountPointReadOnly(mp mount.MountPoint) bool {
```

Comment:
The reason why the fs becomes read-only is the volume corruption, right? Then can we use the CSI fs corruption detection function instead?
BTW, there is one in-progress PR updating that PR. You can have a review as well: https://github.com/longhorn/longhorn-manager/pull/2043

Reply:
According to the user's report
The volume mount became read-only in share-manager but didn't know why.
```
the right thing is :
share-manager-pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922:/tmp # mount |grep pvc
/dev/longhorn/pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922 on /export/pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922 type ext4 (rw,relatime)

the exceptions is:
share-manager-pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922:/tmp # mount |grep pvc
/dev/longhorn/pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922 on /export/pvc-6f4444ac-eab0-4f13-a0b7-5dd915d1f922 type ext4 (ro,relatime)
```
I think the corruption detection function in csi could work since it use `sync()` to test write operation.
Let me try and see if it works

Reply:
After discussing with @derekbit 
Since share-manager runs this function `runHealthCheck` periodically, it is quite heavy to read write test the mount point.
Also we already have 
```
	if err := exec.CommandContext(m.context, "ls", mountPath).Run(); err != nil {
		return false
	}
```
to test the mount point, it will restart the whole share manager pod if it failed to list the mountPath

This function `isMountPointReadOnly` here after the "ls" check is only to test if mount become read-only and remount it without restart the whole pod.
WDYT?

Reply:
> This function isMountPointReadOnly here after the "ls" check is only to test if mount become read-only and remount it without restarting the whole pod.

In this case, does the NFS server still work fine without restarting? If YES, then this PR LGTM

Reply:
Yes, I have tested the `mount -o remount,rw` take effect immediately and wouldn't restart or crash the nfs server
and one can write the data again